### ISSUE TITLE
RAG with Neo4j nodes and relationships as context for query generation

### ISSUE THREAD

**Issue by gurveervirk** [run-llama/llama_index#12614]
### Question Validation

- [X] I have searched both the documentation and discord for an answer.

### Question

I am trying to query my Neo4j graph db using Mistral and nl2graphquery/nl2cypher. I ended up trying RetrieverQueryEngine with KnowledgeGraphRAGRetriever with Neo4jGraphStore as well as the KnowledgeGraphQueryEngine. I ran into an error that was mentioned in [this issue](https://github.com/run-llama/llama_index/issues/11034) and tried specifying graph_query_synthesis_prompt using the respective legacy template but the quality of the generated query was not up to the mark and the query did not execute correctly and didn't retrieve any response from Neo4j.

My doubt is regarding the current prompt template setup. _Does this entire setup incorporate retrieving all the nodes and relationships and comparing them with the generated keywords and embeddings to generate the appropriate query? If not, why?_

I have gone through the llama-index documentation and respective code in github and couldn't find a concrete answer.

**Comment by logan-markewich** [comment]
`Does this entire setup incorporate retrieving all the nodes and relationships and comparing them with the generated keywords and embeddings to generate the appropriate query?`

No. The KnowledgeGraphRAGQueryEngine will a) generate keywords/synoyms based on the query text alone, and optionally b) generate a text2cypher statement, using some schema from the knowledge graph DB.

**Comment by gurveervirk** [comment]
Thanks for the responses!

Currently, the prompt might include a keyword that does exist in the Neo4j either as a node or a relationship, but the representation might not be same, specifically in terms of case (lowercase, uppercase etc) or maybe those keywords are classes or functions in programming languages, which would not be captured appropriately.

I have encountered the above problem in my initial testing which is leading to a poor quality of response. Is it possible to incorporate comparison of the existing nodes and relationships in the graph with the prompt, which might narrow down the possible queries and improve response quality?

I can provide an example of the problem I am encountering as a colab notebook, if need be.

**Comment by kuguadawang12138** [comment]
I feel that there is something wrong with this search engine itself, especially for Neo4j.I've also tried to connect with KG-RAG and all the returns are Empty Response.

**Comment by logan-markewich** [comment]
The llm generates keywords. If no keywords match an triplets (like exactly match) then nothing gets returned.

More than happy to have a pr if something is wrong, but if you stepped through the code in a debugger, you'd see what is happening

**Comment by gurveervirk** [comment]
Yeah, I tested that in colab and found that only after a more context or domain oriented prompt template was created, that the llm was able to generate at least 1 keyword that I actually wanted.

I'll try to compare the nodes and relationships retrieved from Neo4j with the question and then find the keywords and see if that works better.

> The llm generates keywords. If no keywords match an triplets (like exactly match) then nothing gets returned.
> 
> More than happy to have a pr if something is wrong, but if you stepped through the code in a debugger, you'd see what is happening

**Comment by gurveervirk** [comment]
I have tested the comparison with an embedding model for simple single-part question, and it works pretty well. However, I also want to provide the documents as a context during retrieval and response. 

How do I provide the documents and the previously generated Neo4j graph_store as input to the KnowledgeGraphIndex? Is there a way to automatically rehash the documents to the values in the graph_store?

> Yeah, I tested that in colab and found that only after a more context or domain oriented prompt template was created, that the llm was able to generate at least 1 keyword that I actually wanted.
> 
> I'll try to compare the nodes and relationships retrieved from Neo4j with the question and then find the keywords and see if that works better.
> 
> > The llm generates keywords. If no keywords match an triplets (like exactly match) then nothing gets returned.
> > More than happy to have a pr if something is wrong, but if you stepped through the code in a debugger, you'd see what is happening

**Comment by kuguadawang12138** [comment]
New to llama_index. @gurveervirk Can I see how you tested it？Thank you so much.

**Comment by gurveervirk** [comment]
@dosu-bot How do I reinitialize a knowledgegraphindex using my pre-built neo4j graph_store and my documents? I want to convert the index to a query_engine and use the documents for querying also. And does the retrieverqueryengine use the nodes and relationships in any manner?

**Comment by gurveervirk** [comment]
@dosu-bot 
I am able to create a KGI using the following which only uses my Neo4j graph store which was built using some documents:

graph_store = Neo4jGraphStore(
    username=username,
    password=password,
    url=url,
    database=database,
)
storage_context = StorageContext.from_defaults(graph_store=graph_store)
index = KnowledgeGraphIndex([], storage_context=storage_context, graph_store=graph_store)

However, I also want to pass the documents to it in place of [], without it recreating the graph. Meaning, I want the index to use my graph as well as the docs that were used to create the graph as context for the llm during querying, somewhat similar to how a KGI, just after being created using documents, can be converted to a query_engine using as_query_engine, but at a later time.

**Comment by gurveervirk** [comment]
I was going through the github code for KGTableRetriever and have a doubt in this code:
```
def _retrieve(
        self,
        query_bundle: QueryBundle,
    ) -> List[NodeWithScore]:
        """Get nodes for response."""
        node_visited = set()
        keywords = self._get_keywords(query_bundle.query_str)
        if self._verbose:
            print_text(f"Extracted keywords: {keywords}\n", color="green")
        rel_texts = []
        cur_rel_map = {}
        chunk_indices_count: Dict[str, int] = defaultdict(int)
        if self._retriever_mode != KGRetrieverMode.EMBEDDING:
            for keyword in keywords:
                subjs = {keyword}
                node_ids = self._index_struct.search_node_by_keyword(keyword)
```
I believe this is where the keywords are being generated by the llm. Why is this being done? Aren't embeddings enough (assuming embeddings are being used)?

After testing with 2 different (domain) graphs, this keyword generation does little good, more often than not leads the model down the wrong sub-graph. It also wastes time while querying.

If embeddings are not being used and no documents and docstore is being used for KGI creation (like at a later time), then only this keyword generation is being used for responses (which is like shooting in the dark).

@dosu-bot, @logan-markewich Is there a way to create embeddings from the graph_store during initialization of KGI or KnowledgeGraphQueryEngine and use those for response generation in some way (not using load_index_from_storage)?

> The llm generates keywords. If no keywords match an triplets (like exactly match) then nothing gets returned.
> 
> More than happy to have a pr if something is wrong, but if you stepped through the code in a debugger, you'd see what is happening

**Comment by gurveervirk** [comment]
@dosu-bot How do I test a branch of a fork of this repo for changes?
